### DISCUSSION TITLE
stream_write 与 stream_read 接口的问题

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#207] [Q&A (问题和回答)] [answered]
stream_write 与 stream_read 接口有几个问题请教一下大家：
1，每次stream_write调用完成后，都必须 quic_endpoint_process_connections 一下吗（验证了如果write后，没有process connection，就会变idle，write的数据实际没有发出去）？ 或者可以多次stream_write直到写失败后，再quic_endpoint_process_connections 一下？
2，stream_write 的返回值，是否可能大于0且小于目标长度，这种情况下，可以立马再写一次，还是需要等 writeable 回调？
3，stream_read的返回值如果是大于0且小于目标长度，是否立马再读一次，还是需要等readable回调？
4，quic_stream_wantwrite 在什么情况下需要调用？
@iyangsj

**Comment by glime-ai** [comment]
1. 不是每次`stream_write`调用完成后都必须调用`quic_endpoint_process_connections`。你可以进行多次`stream_write`操作，然后再统一调用`quic_endpoint_process_connections`来处理连接。这样做可以提高效率，减少处理次数。但是，确保在适当的时候调用`quic_endpoint_process_connections`是很重要的，以确保数据被发送出去。

2. `stream_write`的返回值确实可能大于0且小于目标长度。这表示只有部分数据被写入。在这种情况下，你可以选择立即再次尝试写入剩余的数据，但这可能会导致EAGAIN错误（如果流不可写）。通常，最好的做法是等待`writable`回调，然后再继续写入剩余的数据，这样可以避免不必要的重试和潜在的性能问题。

3. 对于`stream_read`，如果返回值大于0且小于目标长度，这意味着有一些数据可读，但没有达到请求的长度。你可以选择立即再次尝试读取更多数据，但如果没有更多数据可读，这可能会返回0。通常，等待`readable`回调是一个更好的选择，因为它会在有更多数据可读时通知你。

4. `quic_stream_wantwrite`通常在你想要被通知流可写时调用。例如，如果你在尝试写入数据时收到EAGAIN错误，你可以使用`quic_stream_wantwrite`来请求一个`writable`回调，当流再次可写时，你将被通知，然后可以继续写入数据。

请注意，具体实现细节可能因不同的QUIC库而异，建议参考你所使用的库的文档或源代码以获取最准确的信息。

**Comment by iyangsj** [comment, marked as answer]
1. 可以多次执行写操作后，再执行后置处理 quic_endpoint_process_connections
建议仅在回调函数中（例如writable回调中）执行写操作，这样不必每次显式调用后置处理；同时库会在批量读写操作完成后，才执行后置处理，效率更高

2. stream_write有可能实际写入数据，少于提供的数据，应等待 writeable回调再次尝试写入

3. 同上，应等待readable回调再次尝试读取

4. quic_stream_wantwrite用于暂时关闭或启用可写事件通知
例如：当QUIC流可写而应用层暂时无数据需要写入时，可以暂时屏蔽可写事件，避免无用的回调通知

**Reply by stopandjoin** [reply]
@iyangsj 谢谢！再帮忙确认一下， 
1，writeable 事件会在什么情况下发出， 在上次写入未完成时后面才可能触发吗，写入完成了就不触发？  
2，如果quic_stream_wantwrite 没有设置为false，且并没有调用stream write的情况下， writeable事件会持续重复触发？

**Reply by iyangsj** [reply]
writable事件是水平触发，当流缓冲区用剩余空间时，每个处理周期都会触发

**Reply by stopandjoin** [reply]
ok，明白了。
